### PR TITLE
zvol rework

### PR DESCRIPTION
**Pull request by ryao**
The Illumos taskq import that I have been developing is incompatible with the current zvol code due to how Linux handles interrupts. This refactors the zvol code to stop processing IOs in interrupts. See the commit message of ryao/zfs@1d22dbdf386c3393b1777ba5e06f9de7c368f624 for a full description of the benefits. To summarize, we see the following benefits from this change:

1. No per zvol spinlocks.
2. No redundant IO elevator processing.
3. Interrupts are disabled only when actually necessary.
4. No redispatching of IOs under heavy load.
5. Linux's page out routines will properly block.
6. Many autotools checks become obsolete.

In addition, this pull request includes ryao/zfs@4ed7d715f5221c7ca00e22df9123b6bb4e08c700, which is a fix for a potentially rare deadlock in ZIO processing that became readily reproducible in the presence of this change.

### CODE REVIEW

**Comment by sempervictus** [comment]
@olw2005: so far so good, still seeing some strange rcu_sched lockups on ZVOLs, but have been using the kmem rework code which got merged yesterday on most of the loaded hosts instead of this and my usual stack (ABD, etc).

@ryao: could you please refresh this stack against master? I'd like to reintroduce these changes but am seeing merge conflicts in module/zfs/zvol.c among others (though these look like some significant effort would be required to resolve without on this end). Thanks as always

**Comment by ryao** [comment]
torvalds/linux@3ffb52e73b47d6ad86b645942ff49035efdbcd31explains what is wrong on 2.6.32. Originally, we were supposed to use `BIO_RW_DISCARD`, but at some point after 2.6.32, this was changed to `REQ_DISCARD`. `REQ_DISCARD` existed on 2.6.32, but was not used for this. The consequence is that  the patches modules would build, but they would not perform discard when asked. This should be relatively easy to fix with an autotools check.

**Comment by ryao** [comment]
My curiosity got the better of me. I wrote a fix for the 2.6.32 issue and pushed it for buildbot testing. If all goes well, this should be safe to merge. I do not recommend trying this in production until the build bot results are back because it has not been subject to my own regression tests.

File: module/zfs/zvol.c [hunk 1]
```diff
@@ -608,69 +602,65 @@ zvol_write(void *arg)
...
-
-	if (start >= end) {
-		error = 0;
-		goto out;
+#ifdef REQ_SECURE
+	if (!(bio->bi_rw & REQ_SECURE))
+		start = P2ROUNDUP(start, zv->zv_volblocksize);
+		end = P2ALIGN(end, zv->zv_volblocksize);
 	}
+#ifdef
```
Comment (sempervictus) [inline module/zfs/zvol.c]: I think this is supposed to be #endif for REQ_SECURE, and GCC seems to agree that this is a problem. Strange thing is, the compile-time error only popped up during the DKMS build on one host (several others built, which is now slightly concerning). Everything runs the same kernel, not sure why this one tripped it.

**Comment by behlendorf** [comment]
Bumping to 0.6.5, if we can get it done for 0.6.4 that would be great but it shouldn't hold up the release.

**Comment by sempervictus** [comment]
initial testing is quite encouraging, though i'm still baffled by what is now a consistent build failure on our patch stack - seems this doesnt play well with the large block and ZIO optimization PRs ( #2865 and #3063 ), which really means it should never have built in the first place on the systems where it does work.

iSCSI and local ZVOL backed random IO load from multiple VMs (including running on-host for the local tests with all the contention involved) shows ~15% overall drop in IOWait from the same workload. When running random greps through VM FS' searching for strings which will never match, the reduction in wait is closer to 30%, which seems a bit odd, but i'm not complaining.

Thank you @ryao, i'll get this running in a few test hosts before i try to back any of our production SCST hosts with this, but i highly encourage anyone with resources to put this through its paces. Last time we ran this patch stack we ended up seeing issues pretty late in the game - like on resilvering and other unusual operations, so thorough testing is highly recommended.

**Comment by devonblzx** [comment]
I don't have much experience with git.  Can someone guide me on how I can test the latest commit (2.6.32 fix)?  I have a few CentOS 6.6 dev boxes.

**Comment by FransUrbo** [comment]
@ryao fix:

```
diff --git a/module/zfs/zvol.c b/module/zfs/zvol.c
index e182550..1e8ba38 100644
--- a/module/zfs/zvol.c
+++ b/module/zfs/zvol.c
@@ -641,11 +641,11 @@ zvol_discard(struct bio *bio)
         * 2.6.35) will not receive this optimization.
         */
 #ifdef REQ_SECURE
-       if (!(bio->bi_rw & REQ_SECURE))
+       if (!(bio->bi_rw & REQ_SECURE)) {
                start = P2ROUNDUP(start, zv->zv_volblocksize);
                end = P2ALIGN(end, zv->zv_volblocksize);
        }
-#ifdef
+#endif

        if (start >= end)
                return (0);
```

**Comment by kernelOfTruth** [comment]
@ryao any news or changes on this ?

**Comment by kpande** [comment]
from what i've heard, @ryao is no longer using zvol at his place of work so it'll be a while before this is looked at again. I have observed issues with this patchset. It'll probably be reworked after his API changes are accepted and merged.

**Comment by behlendorf** [comment]
Yes, this change is definitely going to need to be updated and have some additional review and discussion.  @ryao I'm going to close this out for now, please open a new pull request when you have some time to revisit this.

**Comment by ryao** [comment]
@kpande Are these issues on recent kernels or just 2.6.32?

**Comment by ryao** [comment]
@kpande Are these issues on recent kernels or just 2.6.32?

@FransUrbo I just saw your comment now. Thanks for catching that. It was a last ditch attempt to improve this before I put it on the back burner. I just rebased this on head and pushed it to my branch. I haven't had a chance to do testing though.

**Comment by kpande** [comment]
@ryao modern kernels (3.2.0, 3.4.69, 3.17.7)

**Comment by phlogisto** [comment]
@ryao in your patch at this part in zvol.c:
+#ifdef REQ_SECURE
+	if (!(bio->bi_rw & REQ_SECURE))
+		start = P2ROUNDUP(start, zv->zv_volblocksize);
+		end = P2ALIGN(end, zv->zv_volblocksize);
+	}
+#ifdef

think you are missing '{' at start of the block and '#endif' instead of '#ifdef' at the end.

**Pull request closed without merging**